htp/htp_transaction.c

```diff
@@ -834,8 +834,21 @@ htp_status_t htp_tx_res_process_body_data_ex(htp_tx_t *tx, const void *data, siz
             if (tx->connp->out_decompressor == NULL || tx->connp->out_decompressor->decompress == NULL)
                 return HTP_ERROR;
 
+            struct timeval before, after, duration;
+            gettimeofday(&before, NULL);
             // Send data buffer to the decompressor.
             tx->connp->out_decompressor->decompress(tx->connp->out_decompressor, &d);
+            gettimeofday(&after, NULL);
+            //check for overflows ?
```

Comment:
Did not compile test this, but what about something like
```c
struct timeval start, limit, after;
memset(&limit, 0, sizeof(limit));
gettimeofday(&start, NULL);
timeradd(&start, &tx->connp->out_decompressor->time_spent, &limit);

// Send data buffer to the decompressor.
tx->connp->out_decompressor->decompress(tx->connp->out_decompressor, &d);
gettimeofday(&after, NULL);

if ( timercmp(&after, &limit, >) ) {
 // spent too much time
}
```

Reply:
Ah ! I guess I finally understand our difference in thinking :-)

I propose to cumulate the time spent in _multiple_ calls of `decompress` for one connection
While you measure and limit only the time spent in _one_ call to `decompress`

Hence the general question : what should be the scope of the time measurement ? 
I think it safer to cumulate to avoid some evasion by attacker splitting into smaller chunks...
But I am not sure. What is your argument for one call only ?

Reply:
Actually I missed this about your approach. I was only thinking about a way to not have to worry about the timersub overflow case when time might change.

Reply:
I do agree we should track the cumulative cost.

Reply:
ok : added a sanity check for timersub overflow and kept the cumulative cost